src/contract/HandlerBasedContract.ts

```diff
@@ -200,27 +209,86 @@ export class HandlerBasedContract<State> implements Contract<State> {
 
   async readStateFor(
     sortKey: string,
-    interactions: GQLNodeInterface[]
+    interactions: GQLNodeInterface[],
+    signal?: AbortSignal
   ): Promise<SortKeyCacheResult<EvalStateResult<State>>> {
-    return this.readState(sortKey, undefined, interactions);
+    return this.readState(sortKey, interactions, signal);
+  }
+
+  async readStateBatch(pagesPerBatch = 1, signal: AbortSignal): Promise<SortKeyCacheResult<EvalStateResult<State>>> {
+    if (!this.isRoot()) {
+      throw new Error('readStateBatch is only allowed for root contract calls');
+    }
+    if (pagesPerBatch < 1) {
+      throw new Error('At least one page per batch is required');
+    }
+    if (signal.aborted) {
+      throw new Error('readStateBatch aborted');
+    }
+
+    const contractTxId = this._contractTxId;
+    const { interactionsLoader, stateEvaluator } = this.warp;
+    let cachedState = await stateEvaluator.latestAvailableState<State>(contractTxId);
+
+    const evaluationOptions = {
+      ...this._evaluationOptions,
+      transactionsPagesPerBatch: pagesPerBatch
+    };
+
+    let interactions: GQLNodeInterface[];
+    let batchesLoaded = 0;
+    do {
+      const batchBenchmark = Benchmark.measure();
+      this.logger.debug(`Loading ${++batchesLoaded}`, evaluationOptions);
+      interactions = await interactionsLoader.load(contractTxId, cachedState?.sortKey, undefined, evaluationOptions);
+      if (signal.aborted) {
+        throw new Error('readStateBatch aborted');
+      }
+      if (interactions.length == 0) {
+        break;
+      }
+      this.logger.debug(`Evaluating ${interactions.length} in ${batchesLoaded}`);
+      cachedState = await this.readStateFor(cachedState?.sortKey || genesisSortKey, interactions, signal);
+      if (signal.aborted) {
+        throw new Error('readStateBatch aborted');
+      }
+      this.logger.debug(
+        `Batch ${batchesLoaded} evaluated in ${batchBenchmark.elapsed()} at sortKey ${cachedState.sortKey}`
+      );
+    } while (interactions.length > 0);
+
+    return cachedState;
   }
 
   async viewState<Input, View>(
     input: Input,
     tags: Tags = [],
     transfer: ArTransfer = emptyTransfer,
-    caller?: string
+    caller?: string,
+    signal?: AbortSignal
   ): Promise<InteractionResult<State, View>> {
     this.logger.info('View state for', this._contractTxId);
-    return await this.callContract<Input, View>(input, 'view', caller, undefined, tags, transfer);
+    return await this.callContract<Input, View>(
+      input,
+      'view',
+      caller,
+      undefined,
+      tags,
+      transfer,
+      false,
+      false,
+      true,
+      signal
```

Comment:
We should consider some callContractOptions with default values. 
Not necessarly, but maybe in the future?

Reply:
yeah, in general w should refactor to object params in most of the Contract interface methhod.